firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -119,7 +134,107 @@ public void onFailure(@NonNull @NotNull Exception e) {
+
+    if (checkForUpdateTaskCompletionSource != null
+        && !checkForUpdateTaskCompletionSource.getTask().isComplete()) {
+      checkForUpdateCancellationSource.cancel();
+    }
+
+    checkForUpdateCancellationSource = new CancellationTokenSource();
+    checkForUpdateTaskCompletionSource =
+        new TaskCompletionSource<>(checkForUpdateCancellationSource.getToken());
+    checkForUpdateHandler = HandlerCompat.createAsync(Looper.getMainLooper());
+
+    firebaseInstallationsApi
+        .getId()
+        .addOnSuccessListener(
+            new OnSuccessListener<String>() {
+              @Override
+              public void onSuccess(String fid) {
+                firebaseInstallationsApi
+                    .getToken(true)
+                    .addOnSuccessListener(
+                        new OnSuccessListener<InstallationTokenResult>() {
+                          @Override
+                          public void onSuccess(InstallationTokenResult installationTokenResult) {
+
+                            String authToken = installationTokenResult.getToken();
+                            String appId = firebaseApp.getOptions().getApplicationId();
+                            String apiKey = firebaseApp.getOptions().getApiKey();
+
+                            //make into method parameters
+                            FirebaseAppDistributionTesterApiClient client =
+                                new FirebaseAppDistributionTesterApiClient(fid, appId, apiKey);
+
+                            handleLatestReleaseFromClient(client, authToken);
+                          }
+                        });
+              }
+            })
+        .addOnFailureListener(
+            new OnFailureListener() {
+              @Override
+              public void onFailure(@NonNull Exception e) {
+                setCheckForUpdateTaskCompletionError(
+                    new FirebaseAppDistributionException(
+                        FirebaseAppDistributionException.Status.UNKNOWN));
+              }
+            });
+
+    return checkForUpdateTaskCompletionSource.getTask();
+  }
+
+  private void handleLatestReleaseFromClient(
+      FirebaseAppDistributionTesterApiClient client, String authToken) {
+    checkForUpdateExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              AppDistributionRelease latestRelease = client.fetchLatestRelease(authToken);
+
+              Context context = firebaseApp.getApplicationContext();
+              long currentInstalledVersionCode = getInstalledAppVersionCode(context);
+
+              //move logic outside,
+              checkForUpdateHandler.post(
+                  new Runnable() {
+                    @Override
+                    public void run() {
+                      if (currentInstalledVersionCode
+                          < Long.parseLong(latestRelease.getBuildVersion())) {
+                        // latest release is newer
```

Comment:
Could we change this comment to explain the why behind the logic? (something like: // only return a result if the device can upgrade to the new release without uninstalling)